src/main/java/eutros/multiblocktweaker/crafttweaker/predicate/CTTraceabilityPredicate.java

```diff
@@ -0,0 +1,361 @@
+import java.util.Set;
+import java.util.function.Predicate;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+
+@ZenClass("mods.gregtech.multiblock.CTPredicate")
+@ZenRegister
+public class CTTraceabilityPredicate {
+    TraceabilityPredicate internal;
+
+    public CTTraceabilityPredicate(TraceabilityPredicate internal) {
+        this.internal = internal;
+    }
+
+    public CTTraceabilityPredicate(Predicate<BlockWorldState> predicate, Supplier<BlockInfo[]> candidates) {
+        this(new TraceabilityPredicate(predicate, candidates));
+    }
+
+    @ZenConstructor
+    public CTTraceabilityPredicate(IPredicate predicate, ICandidates candidates) {
+        this(new TraceabilityPredicate((state) -> predicate.test(new MCBlockWorldState(state)),
+                () -> Arrays.stream(candidates.get()).map(IBlockInfo::getInternal).toArray(BlockInfo[]::new)));
+    }
+
+    @ZenConstructor
+    public CTTraceabilityPredicate() {
+        this(new TraceabilityPredicate());
+    }
+
+    /**
+     * can be any block.
+     */
+    @ZenMethod
+    public static CTTraceabilityPredicate ANY() {
+        return new CTTraceabilityPredicate(TraceabilityPredicate.ANY);
+    }
+
+    /**
+     * Only the air block.
+     */
+    @ZenMethod
+    public static CTTraceabilityPredicate AIR() {
+        return new CTTraceabilityPredicate(TraceabilityPredicate.AIR);
+    }
+
+    /**
+     * The Wire Coils block. with the same type. will also write its temperature to the context("coils_temperature").
+     */
+    @ZenMethod
+    public static CTTraceabilityPredicate COILS() {
+        return new CTTraceabilityPredicate(new TraceabilityPredicate(blockWorldState -> {
+            net.minecraft.block.state.IBlockState blockState = blockWorldState.getBlockState();
+            if ((blockState.getBlock() instanceof BlockWireCoil)) {
+                BlockWireCoil blockWireCoil = (BlockWireCoil) blockState.getBlock();
+                BlockWireCoil.CoilType coilType = blockWireCoil.getState(blockState);
+                Object currentCoilType = blockWorldState.getMatchContext().getOrPut("CoilType", coilType);
+                blockWorldState.getMatchContext().getOrPut("coils_temperature", coilType.getCoilTemperature());
+                if (!currentCoilType.toString().equals(coilType.getName())) {
+                    blockWorldState.setError(new PatternStringError("gregtech.multiblock.pattern.error.coils"));
+                    return false;
+                }
+                blockWorldState.getMatchContext().getOrPut("VABlock", new LinkedList<>()).add(blockWorldState.getPos());
+                return true;
+            }
+            return false;
+        }, ()-> ArrayUtils.addAll(
+                Arrays.stream(BlockWireCoil.CoilType.values()).map(type->new BlockInfo(MetaBlocks.WIRE_COIL.getState(type), null)).toArray(BlockInfo[]::new)))
+                .addTooltips("gregtech.multiblock.pattern.error.coils"));
+    }
```

Comment:
Could these be static fields instead of functions instead, please? First, full caps function names don't make sense, second, these are all constant.

Reply:
ANY and AIR will be canstant later. but the COILS cant, because toolltips are added to separate predicates, and they need to be created dynamically.

Reply:
Can you make the method name be `getCoils`, then? Also that really sounds like an issue with CEu... Also, if this is exactly the same as `TraceabilityPredicate#HEATING_COILS.get()` then why not just use that?

Reply:
because we cant get the temperature from context with the CEu one. CEu didnt write the temperture of the coils into the the context.

Reply:
Okay, can you please just make sure none of the method names are all caps, though? `getAny`/`getAir`/`getCoils` instead for the static methods.